README
NAME
    WWW::Finger - get useful data from e-mail addresses

SYNOPSIS
      use WWW::Finger;
      my $finger = WWW::Finger->new("joe@example.com");
      if (defined $finger)
      {
        print $finger->name . "\n";
      }

DESCRIPTION
    This module is *not* an implementation of the finger protocol (RFC
    1288). Use Net::Finger for that. Instead it is a set of implementations
    of *other* methods for getting information from an e-mail address, or
    e-mail like identifier. This package includes four such implementations,
    and it's pretty easy to create your own additional implementations:

    *       WebFinger

    *       Fingerpoint

    *       MetaCPAN API for cpan.org addresses

    *       Unnamed finger protocol described on bitworking.org

  Constructor
    *       "new"

              $finger = WWW::Finger->new($identifier);

            Creates a WWW::Finger object for a particular identifier. Will
            return undef if no implemetation is able to handle the
            identifier

  Object Methods
    Any of these methods can return undef if the appropriate information is
    not available. The "name", "mbox", "homepage", "weblog", "image" and
    "key" methods work in both scalar and list context. Depending on which
    implementation was used by "WWW::Finger->new", the object may also have
    additional methods. Consult the documentation of the various
    implementations for details.

    "name"
        The person's name (or handle/nickname).

    "mbox"
        The person's e-mail address (including "mailto:").

    "homepage"
        The person's personal homepage.

    "weblog"
        The person's blog. (There may be some overlap with "homepage".)

    "image"
        An avatar, photo or other image depicting the person.

    "key"
        The URL of the person's GPG/PGP public key.

    "webid"
        A URI uniquely identifying the person. See
        <http://esw.w3.org/topic/WebID>.

    "endpoint"
        A SPARQL Protocol endpoint which may provide additional data about
        the person. (See RDF::Query::Client.)

    "graph"
        An RDF::Trine::Model object holding data about the person. (See
        RDF::Trine.)

SEE ALSO
    Net::Finger.

    <http://code.google.com/p/webfinger/>.

    <http://buzzword.org.uk/2009/fingerpoint/spec>.

    <http://www.perlrdf.org/>.

    fingerw.

AUTHOR
    Toby Inkster, <tobyink@cpan.org>

COPYRIGHT AND LICENCE
    Copyright (C) 2009-2011 by Toby Inkster

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
WWW-Finger
==========

Created:      2009-12-08
Home page:    <https://metacpan.org/release/WWW-Finger>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=WWW-Finger>
Maintainer:   Toby Inkster <mailto:tobyink@cpan.org>

0.104  2012-02-23

 - (Bugfix Packaging) Replace UNIVERSAL::AUTHORITY with Object::AUTHORITY
   in META.yml too.
   ++"Florian Schlichting"
 - (Update) s/Digest::SHA1/Digest::SHA/
   ++"Florian Schlichting"

0.103  2011-12-04

 - (Bugfix Packaging) Newer version of Module::Package::RDF.
 - Replace UNIVERSAL::AUTHORITY with Object::AUTHORITY.

0.102  2011-11-27

 - (Packaging) Module::Package::RDF
 - (Update) use 5.010
 - Cleanup.
 - Rewrite of WWW::Finger::CPAN using MetaCPAN API.
 - Use Module::Pluggable instead of homebrew plugin system.

0.101  2010-07-31

 - (Addition) Webfinger and Fingerpoint modules provide a 'get' method to
   easily access additional data.

0.100  2010-07-25

 - (Addition) CPAN module now has webid method.
 - (Update) Use the somewhat more complete (albeit slower) HTTP::LRDD
   module to perform LRDD for Webfinger rather than a home-grown method.

0.09  2010-04-26

 - (Update) Switch Changes file from Aaron Cope's changefile vocab to my
   own DOAP Change Sets vocab.
 - Various cleanups.

0.08  2010-01-06

 - (Bugfix) Documentation improvements.

0.05  2009-12-16

 - (Bugfix) Mini bug-fixes.
 - (Update) use XRD::Parser 0.04.

0.04  2009-12-16

 - (Addition) Bundle a command-line query tool.
 - (Bugfix) Fingerpoint bugfix in ->graph().
 - (Bugfix) Unimplemented methods return empty arrays properly.
 - Change default scheme for Webfinger from 'acct://' to 'acct:'.

0.03  2009-12-15

 - (Addition) WWW::Finger::Webfinger now provides an ->endpoint method.
 - (Bugfix) Fixed package signature.

0.02  2009-12-14

 - (Bugfix) Fixed problems with build.

0.01  2009-12-14  # Initial release


0.07  2009-01-06

 - (Bugfix) Include BitworkingFingerProtocol.pm in MANIFEST.
 - (Update) Update copyright noticed to 2010.

0.06  2009-01-06

 - (Addition) Allow fingerw command-line client to load additional modules.
 - (Addition) Follow rdf:seeAlso and ianalink:describedby links found in
   Webfinger account profiles.
 - (Addition) Support Joe Gregorio's alternative to current Webfinger
   protocol.
 - (Addition) Support semantics of foaf:account and foaf:holdsAccount.
 - (Bugfix) Fix duplicated mbox in WWW::Finger::CPAN.
 - (Update) Improve fingerw's debugging (verbose) messages.
 - (Update) Use Turtle for Changes file.
 - Refactor some code from WWW::Finger::Webfinger to allow it to be used
   elsewhere.



Makefile.PL
use strict;
use warnings;

use inc::Module::Install;

license             'perl';

name                'WWW-Finger';
author              'Toby Inkster <tobyink@cpan.org>';
perl_version_from   'lib/WWW/Finger.pm';
version_from        'lib/WWW/Finger.pm';
abstract_from       'lib/WWW/Finger.pm';
readme_from         'lib/WWW/Finger.pm';

test_requires       'Test::More' => '0.61';

requires            'Carp'               => '1.00';
requires            'DateTime'           => 0;
requires            'Digest::MD5'        => '2.00';
requires            'Digest::SHA1'       => '2.00';
requires            'Getopt::Long'       => 0;
requires            'HTTP::Link::Parser' => '0.03';
requires            'LWP::Simple'        => '5.00';
requires            'LWP::UserAgent'     => '5.00';
requires            'Pod::Usage'         => 0;
requires            'RDF::Query'         => '2.200';
requires            'RDF::Query::Client' => '0.02';
requires            'RDF::Trine'         => '0.112';
requires            'XML::LibXML'        => '1.60';
requires            'XRD::Parser'        => '0.03';
requires            'URI'                => '1.30';

install_script 'fingerw';

auto_set_homepage;
auto_set_bugtracker;

include 'Test::Signature';
auto_install;
WriteAll(
	'meta' => 1,
	'sign' => 1,
	);


lib/WWW/Finger.pm
package WWW::Finger;

use 5.010;
use common::sense;
use utf8;

use Object::AUTHORITY 0;
use Carp 0;

use namespace::clean;

use Module::Pluggable
	search_path => [qw/WWW::Finger WWW::FingerX/],
	except      => qr/^WWW::Finger(X?)::_/,
	require     => 1,
	inner       => 0,
	sub_name    => 'plugins',
	;

BEGIN {
	$WWW::Finger::AUTHORITY = 'cpan:TOBYINK';
	$WWW::Finger::VERSION   = '0.103';
}

sub new
{
	my $class      = shift;
	my $identifier = shift;

	my @Modules = 
		sort { $a->speed <=> $b->speed }
		$class->plugins;

	foreach my $module (@Modules)
	{
		my $rv = $module->new($identifier);
		return $rv if defined $rv;
	}
	
	return undef;
}

sub name     { return; }
sub mbox     { return; }
sub key      { return; }
sub image    { return; }
sub homepage { return; }
sub weblog   { return; }
sub endpoint { return undef; }
sub webid    { return undef; }
sub graph    { return undef; }

1;

__END__

=head1 NAME

WWW::Finger - get useful data from e-mail addresses

=head1 SYNOPSIS

  use WWW::Finger;
  my $finger = WWW::Finger->new("joe@example.com");
  if (defined $finger)
  {
    print $finger->name . "\n";
  }

=head1 DESCRIPTION

This module is I<not> an implementation of the finger protocol (RFC 1288).
Use Net::Finger for that. Instead it is a set of implementations of
I<other> methods for getting information from an e-mail address, or e-mail
like identifier. This package includes four such implementations, and it's
pretty easy to create your own additional implementations:

=over 8

=item * WebFinger

=item * Fingerpoint

=item * MetaCPAN API for cpan.org addresses

=item * Unnamed finger protocol described on bitworking.org

=back

=head2 Constructor

=over 8

=item * C<new>

  $finger = WWW::Finger->new($identifier);

Creates a WWW::Finger object for a particular identifier. Will return
undef if no implemetation is able to handle the identifier

=back

=head2 Object Methods

Any of these methods can return undef if the appropriate information
is not available. The C<name>, C<mbox>, C<homepage>, C<weblog>,
C<image> and C<key> methods work in both scalar and list context.
Depending on which implementation was used by C<< WWW::Finger->new >>,
the object may also have additional methods. Consult the
documentation of the various implementations for details.

=over

=item C<name>

The person's name (or handle/nickname).

=item C<mbox>

The person's e-mail address (including "mailto:").

=item C<homepage>

The person's personal homepage.

=item C<weblog>

The person's blog. (There may be some overlap with C<homepage>.)

=item C<image>

An avatar, photo or other image depicting the person.

=item C<key>

The URL of the person's GPG/PGP public key.

=item C<webid>

A URI uniquely identifying the person. See L<http://esw.w3.org/topic/WebID>.

=item C<endpoint>

A SPARQL Protocol endpoint which may provide additional data about the person.
(See L<RDF::Query::Client>.)

=item C<graph>

An RDF::Trine::Model object holding data about the person. (See L<RDF::Trine>.)

=back

=head1 SEE ALSO

L<Net::Finger>.

L<http://code.google.com/p/webfinger/>.

L<http://buzzword.org.uk/2009/fingerpoint/spec>.

L<http://www.perlrdf.org/>.

L<fingerw>.

=head1 AUTHOR

Toby Inkster, E<lt>tobyink@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENCE

Copyright (C) 2009-2011 by Toby Inkster

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 DISCLAIMER OF WARRANTIES

THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.

=cut



lib/WWW/Finger/CPAN.pm
package WWW::Finger::CPAN;

use 5.010;
use common::sense;
use utf8;

use Digest::MD5 0 qw(md5_hex);
use JSON 2.00 qw(from_json);
use LWP::Simple 0 qw(get);

use parent qw(WWW::Finger);

BEGIN {
	$WWW::Finger::CPAN::AUTHORITY = 'cpan:TOBYINK';
	$WWW::Finger::CPAN::VERSION   = '0.103';
}

sub speed { 50 }

sub new
{
	my $class = shift;
	my $ident = shift;
	my $self  = bless {}, $class;

	$ident = "mailto:$ident"
		unless $ident =~ /^[a-z0-9\.\-\+]+:/i;
	$ident = URI->new($ident);
	
	return undef
		unless $ident;
		
	$self->{ident} = $ident;
	
	my ($user, $host) = split /\@/, $self->{ident}->to;
	return undef
		unless lc $host eq 'cpan.org';
	
	return $self;
}

sub name
{
	my $self = shift;
	my $name = $self->metacpan_data->{name};
	
	return wantarray
		? @{ [$name] }
		: $name;
}

sub mbox
{
	my $self = shift;	
	my @e    = @{$self->metacpan_data->{email}};
	
	return wantarray
		? @e
		: $e[0];
}

sub pauseid
{
	my $self = shift;
	my ($user, $host) = split /\@/, uc $self->{ident}->to;

	return wantarray
		? @{[ $user ]}
		: $user;
}

sub cpanpage
{
	my $self = shift;
	my $user = $self->pauseid;
	my $cpanpage = 'https://metacpan.org/author/' . (uc $user) . '/';
	
	return wantarray
		? @{[$cpanpage]}
		: $cpanpage;
}

sub homepage
{
	my $self = shift;
	my @hp   = @{$self->metacpan_data->{website}};
	
	push @hp, scalar $self->cpanpage
		unless grep { $self->cpanpage eq $_ } @hp;
	
	return wantarray
		? @hp
		: $hp[0];
}

sub image
{
	my $self = shift;
	my $md5  = lc md5_hex(lc $self->{ident}->to);
	
	return wantarray
		? @{ ["http://www.gravatar.com/avatar/$md5.jpg"] }
		: "http://www.gravatar.com/avatar/$md5.jpg";
}

sub weblog
{
	my $self = shift;	
	my @blog = map { $_->{url} } @{$self->metacpan_data->{blog}};
	
	return wantarray
		? @blog
		: $blog[0];
}

sub metacpan_data
{
	my $self = shift;
	
	unless ($self->{metacpan_data})
	{
		my $user = $self->pauseid;
		my $uri  = sprintf('http://api.metacpan.org/v0/author/%s', uc $user);
		$self->{metacpan_data} = from_json(get($uri));
	}
	
	$self->{metacpan_data};
}

sub releases
{
	my $self = shift;
	
	unless ($self->{releases})
	{
		my $user = $self->pauseid;
		my $uri  = sprintf('http://api.metacpan.org/v0/release/_search?q=author:%s%%20AND%%20status:latest&size=100', uc $user);
		$self->{releases} = [ map { $_->{_source} } @{ from_json(get($uri))->{hits}->{hits} } ];
	}
	
	return wantarray
		? @{ $self->{releases} }
		: scalar @{ $self->{releases} };
}

sub webid
{
	my $self = shift;
	my $user = $self->pauseid;
	my $webid = 'http://purl.org/NET/cpan-uri/person/' . lc $user;
	
	return wantarray
		? @{[ $webid ]}
		: $webid;
}

1;
__END__

=head1 NAME

WWW::Finger::CPAN - WWW::Finger implementation using MetaCPAN.

=head1 DESCRIPTION

Additional methods (other than standard WWW::Finger):

=over

=item * C<pauseid> - returns the person's PAUSE ID.

=item * C<cpanpage> - returns the person's metacpan.org homepage.

=item * C<metacpan_data> - hashref of interesting data.

=item * C<releases> - list of releases. If called in scalar context, count of releases.

=back

=head1 SEE ALSO

L<WWW::Finger>.

=head1 AUTHOR

Toby Inkster, E<lt>tobyink@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENCE

Copyright (C) 2009-2011 by Toby Inkster

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 DISCLAIMER OF WARRANTIES

THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.

=cut


lib/WWW/Finger/Webfinger.pm
package WWW::Finger::Webfinger;

use 5.010;
use common::sense;
use utf8;

use Carp 0;
use HTTP::LRDD 0.104;
use LWP::UserAgent 0;
use RDF::Query 2.900;
use RDF::Trine 0.135 qw[iri];
use URI 0;
use URI::Escape 0;
use XRD::Parser 0.102;

use parent qw(WWW::Finger::_GenericRDF);

BEGIN {
	$WWW::Finger::Webfinger::AUTHORITY = 'cpan:TOBYINK';
	$WWW::Finger::Webfinger::VERSION   = '0.103';
}

sub speed { 100 }

sub new
{
	my $class = shift;
	my $ident = shift or croak "Need to supply an account address\n";
	my $self  = bless {}, $class;

	$ident = "acct:$ident"
		unless $ident =~ /^[a-z0-9\.\-\+]+:/i;
	$ident = URI->new($ident);
	return undef
		unless $ident->scheme =~ /^(mailto|acct|xmpp)$/;
	$self->{'ident'} = $ident;

	my $lrdd = HTTP::LRDD->new('http://lrdd.net/rel/descriptor', 'describedby', 'lrdd', 'webfinger');
	my @d = $lrdd->discover($ident);
	$self->{'graph'} = $lrdd->process_all($ident);
	
	$self->follow_seeAlso(0);
	
	return undef
		unless $self->{'graph'}->count_statements(iri($ident),undef,undef)
		||     $self->{'graph'}->count_statements(undef,undef,iri($ident));
	
	return $self;
}

sub _simple_sparql
{
	my $self = shift;
	my $where = '';
	foreach my $p (@_)
	{
		$where .= " UNION " if length $where;
		$where .= sprintf('{ <%s> <%s> ?x . } '
				. 'UNION { ?z xrd:alias <%s> ; <%s> ?x . } '
				. 'UNION { ?z <http://xmlns.com/foaf/0.1/account> <%s> ; <%s> ?x . } '
				. 'UNION { ?z <http://xmlns.com/foaf/0.1/holdsAccount> <%s> ; <%s> ?x . }',
			(''.$self->{'ident'}), $p,
			(''.$self->{'ident'}), $p,
			(''.$self->{'ident'}), $p,
			(''.$self->{'ident'}), $p,
			);
	}
	
	my $sparql = "PREFIX xrd: <http://ontologi.es/xrd#> SELECT DISTINCT ?x WHERE { $where }";
	my $query  = RDF::Query->new($sparql);
	my $iter   = $query->execute( $self->{'graph'} );
	my @results;
	
	while (my $row = $iter->next)
	{
		push @results, $row->{'x'}->literal_value
			if $row->{'x'}->is_literal;
		push @results, $row->{'x'}->uri
			if $row->{'x'}->is_resource;
	}
	
	if (wantarray)
	{
		return @results;
	}
	
	if (@results)
	{
		return $results[0];
	}
	
	return;
}

sub webid
{
	return (shift)->SUPER::webid(@_);
}

1;

__END__

=head1 NAME

WWW::Finger::Webfinger - WWW::Finger module for Webfinger

=head1 DESCRIPTION

Webfinger is currently a very unstable specification, with implementation details
changing all the time. Given this instability, it seems prudent to describe the
protocol, as implemented by this package.

Given an e-mail-like identifier, the package will prepend "acct:" to it, assuming that
the identifier doesn't already have a URI scheme. This identifier will now be called
[ident].

The package looks up the host-meta file associated with the host for [ident].
It is assumed to be formatted according to the draft-hammer-hostmeta-05
Internet Draft L<http://tools.ietf.org/html/draft-hammer-hostmeta-05> and
XRD Working Draft 10 <http://www.oasis-open.org/committees/download.php/35274/xrd-1.0-wd10.html>.
Both these drafts are dated 19 November 2009.

A link template will be extracted from the host-meta for the host using either
of the following two relationships: L<http://lrdd.net/rel/descriptor>,
L<http://www.iana.org/assignments/relation/lrdd>. (Neither is prioritised, so
if both exist and have different templates, hilarity will ensue.)

The token "{uri}" in the link template will be replaced with the URL-encoded
version of [ident] to create an account descriptor URI.

The account descriptor URI is fetched via HTTP GET with an Accept header
asking for RDF/XML, Turtle, RDF/JSON or XRD. The result is parsed for account
description data if it has status code 200 (OK).

The following relationships/properties are understood in the account
description:

=over

=item * http://xmlns.com/foaf/0.1/name

=item * http://xmlns.com/foaf/0.1/homepage

=item * http://webfinger.net/rel/profile-page

=item * http://xmlns.com/foaf/0.1/weblog

=item * http://xmlns.com/foaf/0.1/mbox

=item * http://webfinger.net/rel/avatar

=item * http://xmlns.com/foaf/0.1/img

=item * http://xmlns.com/foaf/0.1/depiction

=item * http://ontologi.es/sparql#endpoint

=back

As well as the standard WWW::Finger methods, WWW::Finger::Webfinger provides this
additional method:

=over

=item C<< get($p1, $p2, ...) >>

$p1, $p2 and are RDF predicate URIs, XRD Link@rel values, or XRD Property@type values

  # Returns phone numbers...
  $finger->get('http://xmlns.com/foaf/0.1/phone',
               'http://rdf.data-vocabulary.org/#tel');
  
  # Salmon-style magic keys
  $finger->get('magic-public-key');

=back

=head1 SEE ALSO

L<WWW::Finger>, L<XRD::Parser>, L<HTTP::LRDD>.

L<http://code.google.com/p/webfinger/>.

=head1 AUTHOR

Toby Inkster, E<lt>tobyink@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENCE

Copyright (C) 2009-2011 by Toby Inkster

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 DISCLAIMER OF WARRANTIES

THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.

=cut


lib/WWW/Finger/Fingerpoint.pm
package WWW::Finger::Fingerpoint;

use 5.010;
use common::sense;
use utf8;

use Carp 0;
use Digest::SHA1 0 qw[sha1_hex];
use HTTP::Link::Parser 0.102 qw[:standard];
use LWP::UserAgent 0;
use RDF::Query::Client 0.106;
use RDF::Trine 0.135;
use URI 0;

use parent qw[WWW::Finger];

BEGIN {
	$WWW::Finger::Fingerpoint::AUTHORITY = 'cpan:TOBYINK';
	$WWW::Finger::Fingerpoint::VERSION   = '0.103';
}

use constant rel_fingerpoint => 'http://ontologi.es/sparql#fingerpoint';

sub speed { 90 }

sub new
{
	my $class = shift;
	my $ident = shift or croak "Need to supply an e-mail address\n";
	my $self  = bless {}, $class;
		
	$ident = "mailto:$ident"
		unless $ident =~ /^[a-z0-9\.\-\+]+:/i;
	$ident = URI->new($ident);
	return undef
		unless $ident->scheme eq 'mailto';
	
	$self->{'ident'} = $ident;
	my ($user, $host) = split /\@/, $ident->to;
	
	my $ua = LWP::UserAgent->new;
	$ua->timeout(10);
	$ua->env_proxy;
	
	my $httphost = "http://$host/";
	my $response = $ua->head($httphost);
	return undef
		unless $response->is_success;
	
	my $linkdata = HTTP::Link::Parser::parse_links_to_rdfjson($response);
	my $sparql   = $linkdata->{ $httphost }->{ (rel_fingerpoint) }->[0]->{'value'};

	unless (defined $sparql)
	{
		$response = $ua->get($httphost,
			'Accept' => 'application/xhtml+xml;q=1.0, text/html;q=0.9, */*;q=0.1');
		return undef
			unless $response->is_success;
		if ($response->header('content-type') =~ m`^(text/html|application/xhtml+xml|application/xml|text/xml)`i)
		{
			$sparql = URI->new_abs($1, URI->new($httphost))
				if $response->content =~ m`<[Ll][Ii][Nn][Kk]\s+[Rr][Ee][Ll]="[^"]*http://ontologi\.es/sparql#fingerpoint[^"]*"\s+[Hh][Rr][Ee][Ff]="([^"]+)"\s*/?>`;
		}
	}

	return undef
		unless defined $sparql && length $sparql;
	
	$self->{'endpoint'} = $sparql;
	
	my $sha1 = sha1_hex($ident);
	my $sparql_query = "PREFIX foaf: <http://xmlns.com/foaf/0.1/>
	PREFIX wot: <http://xmlns.com/wot/0.1/>
	SELECT DISTINCT *
	WHERE {
		{
			{ ?person foaf:mbox <$ident> . }
			UNION
			{ ?person foaf:mbox_sha1sum \"$sha1\" . }
			UNION
			{ ?person foaf:account <$ident> . }
			UNION
			{ ?person foaf:holdsAccount <$ident> . }
		}
		OPTIONAL { ?person foaf:name ?name . }
		OPTIONAL { ?person foaf:nick ?nick . }
		OPTIONAL { ?person foaf:homepage ?homepage . }
		OPTIONAL { ?person foaf:mbox ?mbox . }
		OPTIONAL { ?person foaf:weblog ?weblog . }
		OPTIONAL { ?person foaf:img ?image . }
		OPTIONAL { ?k wot:pubkeyAddress ?key ; wot:identity ?person . }
	}";
	my @fields = qw(name homepage mbox weblog image key);
	
	my $query  = RDF::Query::Client->new($sparql_query);
	my $result = $query->execute($self->endpoint, {QueryMethod=>'POST'});
	my $webid;

	BINDING: while (my $binding = $result->next)
	{
		$webid = $binding->{'person'}->uri
			if  $binding->{'person'}
			and $binding->{'person'}->is_resource
			and !defined $webid;
			
		FIELD: foreach my $field (@fields)
		{
			next FIELD unless $binding->{$field};
			
			if ($binding->{$field}->is_resource)
				{ $self->{'data'}->{$field}->{ $binding->{$field}->uri } = 1; }
			elsif ($binding->{$field}->is_literal)
				{ $self->{'data'}->{$field}->{ $binding->{$field}->literal_value } = 1; }
		}
	}
	
	foreach my $field (@fields)
	{
		$self->{'data'}->{$field} = [ keys %{ $self->{'data'}->{$field} } ];
	}
	
	$self->{'webid'} = $webid;
	
	return $self;
}

sub graph
{
	my $self = shift;
	
	unless (defined $self->{'graph'})
	{
		my $ident = $self->{'ident'}.'';
		my $sha1 = sha1_hex($ident);
		my $model = RDF::Trine::Model->new( RDF::Trine::Store->temporary_store );
		my $query  = RDF::Query::Client->new("
			PREFIX foaf: <http://xmlns.com/foaf/0.1/>
			DESCRIBE ?person
			WHERE
			{
				{ ?person foaf:mbox <$ident> . }
				UNION
				{ ?person foaf:mbox_sha1sum \"$sha1\" . }
				UNION
				{ ?person foaf:account <$ident> . }
				UNION
				{ ?person foaf:holdsAccount <$ident> . }
			}");
		my $result = $query->execute($self->endpoint, {QueryMethod=>'POST'});
		if ($result)
		{
			while (my $st = $result->next)
			{
				$model->add_statement($st);
			}
			$self->{'graph'} = $model;
		}		
	}
	
	return $self->{'graph'};
}

sub get
{
	my ($self, @params) = @_;
	return WWW::Finger::_GenericRDF::_simple_sparql(
		$self,
		{use_endpoint=>1},
		map { HTTP::Link::Parser::relationship_uri($_) } @params );
}

sub endpoint
{
	my $self = shift;
	return $self->{'endpoint'};
}

sub webid
{
	my $self = shift;
	return $self->{'webid'};
}

sub _data
{
	my $self = shift;
	my $k    = shift;
	if (wantarray)
	{
		return @{ $self->{'data'}->{$k} }
			if defined $self->{'data'}->{$k};
	}
	else
	{
		return $self->{'data'}->{$k}->[0]
			if defined $self->{'data'}->{$k}->[0];
	}
	return undef;
}

sub name
{
	my $self = shift;
	return $self->_data('name');
}

sub nick
{
	my $self = shift;
	return $self->_data('nick');
}

sub mbox
{
	my $self = shift;
	return $self->_data('mbox');
}

sub image
{
	my $self = shift;
	return $self->_data('image');
}

sub homepage
{
	my $self = shift;
	return $self->_data('homepage');
}

sub weblog
{
	my $self = shift;
	return $self->_data('weblog');
}

sub key
{
	my $self = shift;
	return $self->_data('key');
}

1;

__END__

=head1 NAME

WWW::Finger::Fingerpoint - Investigate E-mail Addresses using Fingerpoint

=head1 SYNOPSIS

  ## Using WWW::Finger
  
  use WWW::Finger;
  
  my $finger = WWW::Finger->new("joe@example.com");
  
  if ($finger)
  {
    if ($finger->isa('WWW::Finger::Fingerpoint'))
    {
      print "WWW::Finger used WWW::Fingerpoint\n";
    }
    print $finger->name . "\n";  # print person's name.
 }

  ## Using WWW::Finger::Fingerpoint directly
  
  use RDF::Query::Client;
  use WWW::Finger::Fingerpoint;
  
  my $fingerpoint = WWW::Finger::Fingerpoint->new("joe@example.com");
  
  if ($fingerpoint->webid)
  {
    my $sparql  = sprintf(
      "SELECT * WHERE {<%s> <http://xmlns.com/foaf/0.1/homepage> ?page.}",
      $fingerpoint->webid);
    my $query   = RDF::Query::Client->new($sparql);
    my $results = $query->execute($fingerpoint->endpoint);
	 while (my $row = $results->next)
    {
      print "Found page: " . $row->{'page'}->uri . "\n";
    }
  }
  
=head1 DESCRIPTION

As well as the standard WWW::Finger methods, WWW::Finger::Fingerpoint provides this
additional method:

=over

=item C<< get($p1, $p2, ...) >>

$p1, $p2 and are RDF predicate URIs. Returns a list of values which are non-bnode
objects of triples where the predicate URI is one of the parameters and the 
subject URI is the person/agent fingered.

  # Returns phone numbers...
  $finger->get('http://xmlns.com/foaf/0.1/phone',
               'http://rdf.data-vocabulary.org/#tel');

=back

=head1 SEE ALSO

L<WWW::Finger>.

L<RDF::Query::Client>, L<RDF::Trine>.

L<http://buzzword.org.uk/2009/fingerpoint/spec>.

L<http://www.perlrdf.org/>.

=head1 AUTHOR

Toby Inkster, E<lt>tobyink@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENCE

Copyright (C) 2009-2011 by Toby Inkster

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 DISCLAIMER OF WARRANTIES

THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.

=cut
